File: core/src/main/java/com/neqsoft/reso/device/DeviceFragment.java
package com.neqsoft.reso.device;

import android.content.Context;
import android.os.Bundle;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.neqsoft.reso.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class DeviceFragment extends Fragment {

  private Display display;
  @Nullable
  private DeviceViewModel deviceViewModel;
  @Nullable
  private DeviceRecyclerAdapter deviceRecyclerAdapter;

  @Override
  public void onAttach(@NonNull Context context) {
    super.onAttach(context);
    if (getActivity() != null)
      display = getActivity().getWindowManager().getDefaultDisplay();
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    deviceViewModel = new ViewModelProvider(this).get(DeviceViewModel.class);
  }

  @Nullable
  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.fragment_device, container, false);
    bind(view);
    if (deviceViewModel != null)
      deviceViewModel.getDeviceInfo(display).observe(getViewLifecycleOwner(), this::display);
    return view;
  }

  private void bind(final View view) {
    RecyclerView recyclerView = view.findViewById(R.id.deviceRv);
    recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    deviceRecyclerAdapter = new DeviceRecyclerAdapter();
    recyclerView.setAdapter(deviceRecyclerAdapter);
  }

  private void display(@Nullable Device device) {
    if (deviceRecyclerAdapter != null)
      deviceRecyclerAdapter.submit(device);
  }
}

File: core/src/main/java/com/neqsoft/reso/device/DeviceFragment.kt
package com.neqsoft.reso.device

import android.content.Context
import android.os.Bundle
import android.view.Display
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.neqsoft.reso.R

class DeviceFragment : Fragment() {

    private var display: Display? = null
    private var deviceViewModel: DeviceViewModel? = null
    private var deviceRecyclerAdapter: DeviceRecyclerAdapter? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        display = activity?.windowManager?.defaultDisplay
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        deviceViewModel = ViewModelProvider(this).get(DeviceViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_device, container, false)
        bind(view)
        val device = deviceViewModel?.getDeviceData(display)
        display(device)
        return view
    }

    private fun bind(view: View) {
        val recyclerView: RecyclerView = view.findViewById(R.id.deviceRv)
        recyclerView.layoutManager = LinearLayoutManager(context)
        deviceRecyclerAdapter = DeviceRecyclerAdapter()
        recyclerView.adapter = deviceRecyclerAdapter
    }

    private fun display(device: Device?) {
        deviceRecyclerAdapter?.submit(device)
    }
}